### PR TITLE
Fix module dict key ordering

### LINKED ISSUE #40227: ModuleDict documentation misleadingly describes ordering of keys
## 📚 Documentation

The current documentation says it's an "ordered dictionary that respects the order of insertion", but it can, in fact, destroy its order. For example: 

image

We should update our documentation to clarify its behavior, especially as users might try to enumerate the dict sequentially. 

cc @jlin27 @albanD @mruberry

### PR DESCRIPTION
**Pull request by yl-to**
fix #40227
Removed the sorting operation both in ModuleDict class, updated the docstring.
Also remove a sort operation in corresponding unit test, which will lead to unit test fail.

BC Note: Python version after 3.6, the plain dict will preserve the order of keys.
example:
For a python 3.6+ user, if he is initial a ModuleDict instance using plain python dict:
{
"b": torch.nn.MaxPool2d(3),
"a": torch.nn.MaxPool2d(3)
}
, he will get a ModuleDict which preserve the order:
ModuleDict(
(b): MaxPool2d(kernel_size=3, stride=3, padding=0, dilation=1, ceil_mode=False)
(a): MaxPool2d(kernel_size=3, stride=3, padding=0, dilation=1, ceil_mode=False)
)

For a python 3.5 user, if we maintain the same input, then the output ModuleDict could be:
ModuleDict(
(a): MaxPool2d(kernel_size=3, stride=3, padding=0, dilation=1, ceil_mode=False)
(b): MaxPool2d(kernel_size=3, stride=3, padding=0, dilation=1, ceil_mode=False)
)

### CODE REVIEW

**Comment by yl-to** [comment]
@albanD @colesbury

**Review by albanD — APPROVED** [review]
All tests are green now. Perfect !

**Comment by yl-to** [comment]
> All tests are green now. Perfect !

@albanD Hi, is there anything I need to do now?

**Comment by albanD** [comment]
No, I am landing this.
The PR will be closed and the bot will comment with the commit hash corresponding to the master merge.

**Comment by yl-to** [comment]
> No, I am landing this.
> The PR will be closed and the bot will comment with the commit hash corresponding to the master merge.

Hi, is there any blockage? It is still not merged. @albanD

**Comment by yl-to** [comment]
> > No, I am landing this.
> > The PR will be closed and the bot will comment with the commit hash corresponding to the master merge.
> 
> Hi, is there any blockage? It is still not merged. @albanD

Hi @albanD , can you check this PR when you are free, thanks!

**Comment by albanD** [comment]
No blockage. It just took a while to get all the internal tests to pass and I wanted to avoid landing over the long week end.
